Shazam.php

```diff
@@ -36,6 +33,145 @@ public function __construct()
+
+            $index = array_search($username, $existing_users);
+            if(isset($index)){
+                unset($existing_users[$index]);
+                $re_index = array_values($existing_users); //necessary to prevent conversion to object upon encoding
+                $this->setProjectSetting(self::JS_USER_KEY, json_encode($re_index));
+            } else {
+                $this->emError("Error, existing users should contain {$username}");
+            }
+        } else {
+            $this->emError("User ". USERID . " attempting to remove dynamic js permissions from ${$username} and failed");
+        }
+    }
+
+
+    /**
+     * Returns user options for the UI table dropdown menu.
+     * @param void
+     * @return $html
+     */
+    public function getUserOptions(){
+        if($this->getSystemSetting('enable-add-user-javascript-permissions')){ //Ensure project setting on config is checked
+            $all_users = REDCap::getUsers();
+            $existing_users = $this->getJavascriptUsers();
+            $filtered = array_diff($all_users, $existing_users);
+            $html = "";
+
+            if(!empty($filtered)) {
+                foreach ($filtered as $key => $val) {
+                    if($val !== USERID) { //User shouldn't be able to add themselves
+                        $html .= "<option value={$val}>{$val}</option>";
+                    }
+                }
+            }
+            return $html;
+        }
+    }
+
+    /**
+     * Returns the rows for the user-js-enabled table
+     * @param void
+     * @return $html
+     */
+    public function renderJSTable(){
+        if($this->getSystemSetting('enable-add-user-javascript-permissions')) {
+            $users = $this->getJavascriptUsers();
+            $html = '';
+            $a = gettype($users);
+            if(!empty($users)){
+                foreach($users as $index => $user){
+                    $html .= "<tr>
+                        <td id='user-{$index}'>{$user}</td>
+                        <td><Button class ='btn btn-xs btn-danger removeUser' id={$index}>Remove</Button></td>
+                        </tr>";
+                }
+            }
+            return $html;
+        }
+    }
+
+
+    /**
+     * Get an email address for a username
+     * @param $user
+     * @return String $q
+     */
+    function getUserEmail($user) {
+        $sql = "SELECT user_email from redcap_user_information where username = '" . db_real_escape_string($user) . "'";
+        $q = $this->query($sql);
+        return db_result($q,0);
+    }
+
+    /**
+     * Send email to user notifying them of access
+     * @param $username
+     * @return void
+     */
+    function sendNotificationEmail($username){
+        $to = $this->getUserEmail($username);
+        $messageBody = self::EMAIL_BODY;
+        $emailResult = REDCap::email($to, self::EMAIL_FROM,'Shazam Javascript permissions granted', $messageBody);
```

Comment:
The body should come from the system settings.   Shouldn't the 'from user' come from the currently logged in user, e.g. USER_ID ?